### PR TITLE
 test updating instances in promiseSpace updates agoricNames

### PR DESCRIPTION
**Pull request by Chris-Hibbert**
refs: #9970

## Description

While preparing a coreEval for the vaults&Auctions release, I found it necessary to update both `instance.produce.auctioneer` and `agoricNamesAdmin.lookupAdmin('instance')`. 

This test shows that updating the former causes the latter to update.

### CODE REVIEW

**Comment by Chris-Hibbert** [comment]
For the purposes of conversation, this test shows that 
```
 await space.instance.produce.auctioneer.reset();
  await space.instance.produce.auctioneer.resolve(auctionInstance2);
```
does indeed update agoricNames.

```
  const instanceLookup = await agoricNames.lookup('instance');
  t.is(await instanceLookup.lookup('auctioneer'), auctionInstance2);
```

File: packages/vats/test/name-hub-published.test.js [hunk 1]
```diff
@@ -58,3 +61,30 @@ test('promise space reserves non-well-known names', async t => {
...
+    Object.keys(agoricNamesReserved),
+  );
+
+  const auctionInstance1 = makeHandle('instance1');
+  const instanceLookup = await agoricNames.lookup('instance');
+
+  // @ts-expect-error storing fake instance
+  space.instance.produce.auctioneer.resolve(auctionInstance1);
+  // @ts-expect-error expecting instance
+  t.is(await space.instance.consume.auctioneer, auctionInstance1);
```
Comment (dckc) [inline packages/vats/test/name-hub-published.test.js:81]: I suggest moving the error to where `auctionInstance1` is declared:

```suggestion
  /** @type { Instance } */
  // @ts-expect-error mock
  const auctionInstance1 = makeHandle('instance1');
  const instanceLookup = await agoricNames.lookup('instance');

  space.instance.produce.auctioneer.resolve(auctionInstance1);
  t.is(await space.instance.consume.auctioneer, auctionInstance1);
```

**Review by dckc — COMMENTED** [review]
This shows that the short version works in at least some cases.

It doesn't hurt to land this as-is, but should we look harder for problems first?

**Comment by Chris-Hibbert** [comment]
> It doesn't hurt to land this as-is, but should we look harder for problems first?

I'd like to look harder, but I don't know what to look for. My current (very tentative) hypothesis is that the promises are taking a while to resolve in the upgrade coreEval.

**Comment by dckc** [comment]
> I'd like to look harder, but I don't know what to look for. My current (very tentative) hypothesis is that the promises are taking a while to resolve in the upgrade coreEval.

We can look there, I think, by controlling which promises resolve when in a test. We might have to introduce a few more moving parts. hm.

**Comment by Chris-Hibbert** [comment]
> While preparing a coreEval for the vaults&Auctions release, I found it necessary to update both instance.produce.auctioneer and agoricNamesAdmin.lookupAdmin('instance').

I now suspect that the issue is a subtle form of [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use). 

When dealing with values in the promise space that are being updated during the same run, you can easily get a value from before the change, even if you think one proposal is supposed to happen after another. Particular with multiple core-evals, the process starts them in parallel, and satisfies `consume` directives eagerly.

Similarly, if you retrieve sub-tables from `agoricNames` (as `agoricNamesAdmin.lookupAdmin('instance')` does) you can easily end up with a copy of the sub-table that won't be updated even if you don't look up a specific name in it until later.

@dckc  would you mind if I closed this?  Is the test worth cleaning up and merging?

**Comment by dckc** [comment]
I don't mind. Go ahead

**Pull request closed without merging**